src/frontend/src/binder/expr/function.rs

```diff
@@ -607,41 +610,179 @@ impl Binder {
+            let offset = self
+                .bind_window_frame_bound_offset(*offset, DataType::Int64)?
+                .into_int64();
+            if offset < 0 {
+                return Err(ErrorCode::InvalidInputSyntax(
+                    "offset in window frame bounds must be non-negative".to_string(),
+                )
+                .into());
+            }
+            Ok(offset as usize)
+        };
+        let mut convert_bound = |bound| -> Result<FrameBound<usize>> {
+            Ok(match bound {
+                WindowFrameBound::CurrentRow => FrameBound::CurrentRow,
+                WindowFrameBound::Preceding(None) => FrameBound::UnboundedPreceding,
+                WindowFrameBound::Preceding(Some(offset)) => {
+                    FrameBound::Preceding(convert_offset(offset)?)
+                }
+                WindowFrameBound::Following(None) => FrameBound::UnboundedFollowing,
+                WindowFrameBound::Following(Some(offset)) => {
+                    FrameBound::Following(convert_offset(offset)?)
+                }
+            })
+        };
+        let start = convert_bound(start)?;
+        let end = if let Some(end_bound) = end {
+            convert_bound(end_bound)?
+        } else {
+            FrameBound::CurrentRow
+        };
+        Ok((start, end))
+    }
+
+    fn bind_window_frame_scalar_impl_bounds(
+        &mut self,
+        start: WindowFrameBound,
+        end: Option<WindowFrameBound>,
+        offset_data_type: &DataType,
+    ) -> Result<(FrameBound<ScalarImpl>, FrameBound<ScalarImpl>)> {
+        let mut convert_bound = |bound| -> Result<FrameBound<_>> {
+            Ok(match bound {
+                WindowFrameBound::CurrentRow => FrameBound::CurrentRow,
+                WindowFrameBound::Preceding(None) => FrameBound::UnboundedPreceding,
+                WindowFrameBound::Preceding(Some(offset)) => FrameBound::Preceding(
+                    self.bind_window_frame_bound_offset(*offset, offset_data_type.clone())?,
+                ),
+                WindowFrameBound::Following(None) => FrameBound::UnboundedFollowing,
+                WindowFrameBound::Following(Some(offset)) => FrameBound::Following(
+                    self.bind_window_frame_bound_offset(*offset, offset_data_type.clone())?,
+                ),
+            })
+        };
+        let start = convert_bound(start)?;
+        let end = if let Some(end_bound) = end {
+            convert_bound(end_bound)?
+        } else {
+            FrameBound::CurrentRow
+        };
+        Ok((start, end))
+    }
+
+    fn bind_window_frame_bound_offset(
+        &mut self,
+        offset: ast::Expr,
+        cast_to: DataType,
+    ) -> Result<ScalarImpl> {
+        let mut offset = self.bind_expr(offset)?;
+        if !offset.is_const() {
+            return Err(ErrorCode::InvalidInputSyntax(
+                "offset in window frame bounds must be constant".to_string(),
```

Comment:
We have a [`ConstEvalRewriter`](https://github.com/risingwavelabs/risingwave/blob/8fdd0babd8f883bae830fd7e7fa4809adeb0712f/src/frontend/src/optimizer/plan_expr_rewriter/const_eval_rewriter.rs#L19). Perhaps it can help here.

Reply:
BTW, I saw we did constant folding at the last step in optimization.

https://github.com/risingwavelabs/risingwave/blob/8fdd0babd8f883bae830fd7e7fa4809adeb0712f/src/frontend/src/optimizer/mod.rs#L399-L400

Shall we do it at the beginning instead? @kwannoel cc. @chenzl25

Reply:
We can use `ConstEvalRewriter` to do constant folding if necessary. Here is the binder phase.

Reply:
Oh, indeed. Here is the binder phase 😄

Reply:
Yep, I do constant folding in binder phase here because I want to bind the expression to a well-formed Rust type `FrameBound<ScalarImpl>`, so that later process can be easy. I guess it does few harm since it should always be simply a constant value/expr.